City-of-Helsinki/hel-fi-drupal-grants
# feat: AU-883: Add tilakomponentti

# [AU-883](https://helsinkisolutionoffice.atlassian.net/browse/AU-883)


Add wf component for handling premise details on application forms.

## What was done


* Add new component for premises
* Add DataType + mapper and service class to support mapping
* Update ATV Schema 

## How to install

* Make sure your instance is up and running on correct branch.
  * `git checkout feature/AU-883-tilakomponentti`
  * `make fresh`
* Run `make drush-cr`

## How to test


This can properly be tested only when form is sort of ready. Now we can see that the component works and data is added to ATV. 

* [ ] Add Grants Premises field to either form
* [ ] See that the field appears and you can fill in & save application to ATV
* [ ] Check from ATV document that the premises parts are added
`"activityInfo": {
                "plannedPremisesArray": [
                    [
                        {
                            "ID": "premiseName",
                            "label": "description",
                            "value": "yurtyurtyu",
                            "valueType": "string"
                        },
                        {
                            "ID": "postCode",
                            "label": "description",
                            "value": "55555",
                            "valueType": "string"
                        },
                        {
                            "ID": "isOwnedByCity",
                            "label": "description",
                            "value": false,
                            "valueType": "bool"
                        }
                    ]
                ]
            }`

Also, a datatype config must be added to a form data where you added the form item:

`
      $info['tilat'] = ListDataDefinition::create('grants_premises')
        ->setLabel('Tilat')
        ->setSetting('jsonPath', [
          'compensation',
          'activityInfo',
          'plannedPremisesArray',
        ])
        ->setSetting('usedFields', [
          'premiseName',
          'postCode',
          'isOwnedByCity',
        ])
        ->setSetting('fullItemValueCallback', [
          'service' => 'grants_premises.service',
          'method' => 'processPremises',
        ])
        ->setSetting('fieldsForApplication', [
          'premiseName',
          'isOwnedByCity',
          'postCode',
        ]);
`

## Designers review





[AU-883]: https://helsinkisolutionoffice.atlassian.net/browse/AU-883?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
